### PR TITLE
Warn that register_post_accumulate_grad_hook receives the param, not the grad

### LINKED ISSUE #187268: `register_post_accumulate_grad_hook` naming/documentation conflicts with `register_hook`, leading to semantic misinterpretation of hook argument and silent training bugs
## Summary

`Tensor.register_post_accumulate_grad_hook` passes the **leaf parameter itself** (`param`) to the hook callback, **not the gradient**. While the docstring clarifies this (`hook(param: Tensor) -> None` and states "this hook operates on the tensor that requires grad and not the grad itself"), the function name / opening sentence "after grad accumulation" / the conventional parameter name `g` all strongly imply the hook receives the gradient. This severely conflicts with the naming conventions of other hooks in the same module (`register_hook` receives `grad`, `register_full_backward_hook` receives `(module, grad_input, grad_output)`).

During development of a CPU-offload optimizer for an LLM training project, we fell into this trap: the hook mistakenly treated `g` as the gradient, cast it → D2H → accumulated into a CPU accumulator. After thousands of training steps, loss completely stagnated. Fixing this bug took several days.

We suggest the official team add a proper gradient‑receiving API, or rename the existing API / add a big warning at the top of the docstring.

## Environment

- PyTorch 2.9.1+cu128
- CUDA 12.8
- Reproduction script does not require GPU

## Minimal Reproduction

```python
import torch

# "post_accumulate_grad" strongly suggests the hook receives the gradient — that's how we understood it in our project.
# In reality, the hook receives the leaf parameter itself; to get the gradient you must read param.grad.

v = torch.tensor([0., 0., 0.], requires_grad=True)

# Assume hook(g) receives the newly computed gradient (same convention as register_hook)
received = []
def buggy_hook(g):
    received.append(g.detach().clone())  # Our project: treat g as gradient → cast, flatten, D2H
    # ...

v.register_post_accumulate_grad_hook(buggy_hook)
v.backward(torch.tensor([1., 2., 3.]))

# Expected: received = tensor([1., 2., 3.])   ("post grad accumulation")
# Actual:   received = tensor([0., 0., 0.])   (this is the old value of v.data; gradient is in v.grad)
print(received)        # tensor([0., 0., 0.])  ← bug manifested
print(v.grad)          # tensor([1., 2., 3.])  ← real gradient is here
```

## Compare to Other Hooks (the inconsistency)

| Hook                                               | Receives                             |
|----------------------------------------------------|--------------------------------------|
| `Tensor.register_hook(hook)`                      | `grad: Tensor`                       |
| `Module.register_full_backward_hook(hook)`        | `(module, grad_input, grad_output)`  |
| `Tensor.register_post_accumulate_grad_hook(hook)` | `param: Tensor`, grad at `param.grad`|

Only this hook passes the "parameter" to the callback, while the function name / opening description suggests it is the gradient.

## Real-World Impact

Our project used `register_post_accumulate_grad_hook` to implement per‑parameter D2H gradient streaming (CPU‑offload optimizer). The hook was written as:

```python
def hook(g):
    src = g.detach().to(bf16).reshape(-1).to("cpu", non_blocking=True)
    _pending.append((s.accum, src))   # s.accum is a CPU pinned grad accumulator
    p.grad = None
```

Intent: `g` is the newly computed gradient for the current microbatch; DMA it to CPU and add to `s.accum`. `flush_pending_grads` syncs at the end of each microbatch with `target.add_(src)`, and `optimizer.step()` reads `s.accum` as the gradient.

Reality: `g` is `p.data` (parameter value in BF16), not the gradient. Each microbatch accumulates the parameter value into `s.accum`. After 16 microbatches, `s.accum = 16 × p.data`. The clipping coefficient is `1.0 / 17263 ≈ 5.8e-5`; the optimizer then applies SGD‑style updates to the parameter values themselves, causing almost no parameter movement. Loss stuck at 12.6.

Diagnostic log (production rank‑0):
```
[diag-step 0 pre] total_norm=1.726e+04 muon: acc_max=7.153e-05 mom_max=0.000e+00 pmax=7.617e-02
                                       adamw: acc_max=6.470e-03 m_max=0.000e+00 v_max=0.000e+00 pmax=6.906e+00
```

`pmax=6.906` matches the initial absolute range of `dt_bias` (`[log(-expm1(-0.001)), log(-expm1(-0.1))] = [-6.91, -2.25]`). `acc_max=6.470e-3 = (6.906 / 16) × 1.5e-2` (post‑clip), perfectly consistent with the hypothesis that "what is being accumulated is `p.data`, not the gradient".

Fix: change the hook to read `p.grad`:
```python
def hook(g):           # g is param, not grad
    if p.grad is None:
        return None
    src = p.grad.detach().to(bf16).reshape(-1).to("cpu", non_blocking=True)
    _pending.append((s.accum, src))
    p.grad = None       # must clear manually; hook does not auto‑clear
    return None
```

## Suggested Fixes (any of)

1. **Add a new API that truly passes the gradient**: `register_post_accumulate_grad_hook_for_grad(hook)` with signature `hook(grad: Tensor) -> Optional[Tensor]`. Same semantic as `register_hook`. Emit a `DeprecationWarning` at the entry point of the existing API, recommending migration. Minimal code change.
2. **Rename the existing API** to `register_post_accumulate_grad_hook_on_param` or `register_post_backward_param_hook`, making the "parameter" semantics obvious from the name.
3. **Update the docstring** with a **DANGER** block at the top: put "WARNING: the argument is the leaf tensor, not the gradient" first, before the example, and emphasise "if you wanted the gradient, use `param.grad`".

Preference for **option 1** – it does not break existing user code, provides an intuitive migration path, and eliminates the root cause for new users.

## Stack Trace / Underlying Behavior

On the C++ side, `AccumulateGrad::call_post_hooks` passes the leaf tensor itself (not its `.grad` field) as `g` to the Python hook. This is by design (to allow users to modify the parameter in‑place with `p.add_(p.grad)`). But given that, the parameter semantics should be made unmistakable from the name.

cc @svekars @sekyondaMeta @AlannaBurke @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93

### PR DESCRIPTION
**Pull request by Ashutosh0x**
Tensor.register_post_accumulate_grad_hook passes the leaf tensor itself to the registered hook, not its gradient. The method name and the 'after grad accumulation' phrasing read as though the hook is handed the gradient -- the same convention as register_hook -- so it is easy to treat the argument as the gradient, cast/offload it, and silently corrupt training. The gradient actually lives in param.grad.

The docstring already mentioned this, but only in a plain paragraph below the signature where it is easy to miss. Promote it to a prominent warning admonition that contrasts with register_hook and points users to param.grad. Documentation only; no behavior change.

Fixes #187268

Test Plan:
Documentation-only change; no runtime behavior is affected. Verified the module still byte-compiles and the existing doctest example is unchanged:

`
python -m py_compile torch/_tensor.py
`

This change was authored with the assistance of an AI coding assistant.

### CODE REVIEW

**Comment by soulitzer** [comment]
I closed the corresponding issue
